Color Picker: Re-instate debounce and controlled value to fix issue with gradient picker

andrewserong opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/36066

## Description
<!-- Please describe what you have changed or added -->

Fixes part of #36899, possibly introduced in (or related to #35670)

When editing a gradient (e.g. attached to a background for a Group block), if you open up the color picker and click and drag quickly, prior to this PR, the picker throws an error (reported in #36899).

By reinstating the `debounce` in the color picker, when used as a controlled component (as in the case of the gradient picker) we avoid recursively updating state and causing the block and picker to throw an error.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. Prior to applying this PR, add a Group block to a post, while running a theme that supports gradients.
2. Add a gradient as a background.
3. Click on one of the existing control points in the gradient to open up a controlled instance of the color picker.
4. Click and drag the color picker quickly and notice that it throws an error.
5. Apply this PR, and repeat the above steps — there should be no error thrown.

Test that this doesn't re-introduce any issues that #35670 resolved.

## Screenshots <!-- if applicable -->

### Before



https://user-images.githubusercontent.com/14988353/143816294-0c75a5bd-2579-4a93-9003-eb04fb07110c.mp4


https://user-images.githubusercontent.com/14988353/143816158-5a8eab35-843a-4297-aa45-a3814340a3f3.mp4



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
Color Picker: Re-instate debounce and controlled value to fix issue with gradient picker

aaronrobertshaw reviewed (commented):
Thanks for putting this together @andrewserong 👍 

I believe this also fixes https://github.com/WordPress/gutenberg/issues/36066 so I've added it to the PR description. 

✅ Unit tests pass
✅ Original issue could be replicated
✅ After applying this PR, no error is thrown
✅ Reinstated code matches what was removed
✅ No severe unexpected movements to color selection indicator (see comment below)

When testing that this doesn't re-introduce any issues #35670 was aimed at fixing, I did notice a really small jump to the color selection indicator sometimes while slowly dragging it. I don't believe its significant, it's definitely better than the picker crashing. It also could be tired eyes 👀.

https://user-images.githubusercontent.com/60436221/143824394-16721af1-c055-4e2a-8573-64bbbde56169.mp4

In general, this is looking pretty good to me. I'd be inclined to approve it however it would probably be best to wait upon Jorge's review.

jorgefilipecosta commented on packages/components/src/color-picker/component.tsx:
```diff
@@ -57,23 +59,31 @@ const ColorPicker = (
+
 	// Use a safe default value for the color and remove the possibility of `undefined`.
 	const safeColordColor = useMemo( () => {
 		return color ? colord( color ) : colord( defaultValue );
```
Given that useControlledValue, is already taking into account defaultValue, I guess we can simplify this to `return colord( color )`.

andrewserong replied:
Good suggestion, thanks! I've simplified the return statement here and moved the inline comment to above `useControlledValue` as that's where the default value logic is now happening.

jorgefilipecosta reviewed (approved):
Previously we had a bug on our code and on component rerenders even if the color did not change we were passing a different object to the react colorful component, and that caused a similar crash to the one happening here and unexpected popover closes. That issue was solved and now we use a string version of react-colorful and we only pass a different string if, in fact, a color change happened. But as we can see it is still possible to crash the component. This crash seems internal to react-colorful (we can see on the stack trace that the crash in one of the component hooks) we are just passing a string to it, and I checked using the debugger we only pass a different string when the color changes so I think the ideal fix would be upstream in react-colorful as it seems the component has a bug. That said this seems to fix the issue for us so I think we should merge this PR.

This PR does not revert the fix on https://github.com/WordPress/gutenberg/pull/35670, the fix there was not to make sure we don't pass a different value to react-colorful on rerenders. It just seemed that the fix would allow removing debouncing and that is not the case.

Thank you for fixing this issue @andrewserong.

ciampo reviewed (approved):
Echoing @aaronrobertshaw and @jorgefilipecosta , code changes LGTM 🚀 

Thank you so much for looking into this fix, @andrewserong !

PS: before merging, could you **add a CHANGELOG entry** for this PR (and maybe add also an entry for #35670, as explained [here](https://github.com/WordPress/gutenberg/pull/35670#discussion_r756348528)) ? Thank you!

andrewserong committed:
Add changelog, update readme, simplify safeColordColor return value

andrewserong commented:
Thanks for the reviews and feedback, everyone! I've added changelog entries, and updated the Readme while I was at it to include the `defaultValue` based on [your suggestion](https://github.com/WordPress/gutenberg/pull/35670#discussion_r756348528) in the other PR @ciampo.

I'll merge this in once the tests pass, but let me know if any other follow-ups are required and I can open up another PR.

andrewserong committed:
Remove defaultValue from useMemo dependency array since it is no longer used in the useMemo

andrewserong commented:
I didn't get a chance to merge this in today (we were debugging failing e2e test failures separately in #36982), but 🤞 we can merge this one in tomorrow.
